gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/FilterStep.java

```diff
@@ -34,9 +36,22 @@ public FilterStep(final Traversal.Admin traversal) {
     @Override
     protected Traverser.Admin<S> processNextStart() {
         while (true) {
-            final Traverser.Admin<S> traverser = this.starts.next();
-            if (this.filter(traverser))
-                return traverser;
+            try {
+                final Traverser.Admin<S> traverser = this.starts.next();
+                if (this.filter(traverser))
+                    return traverser;
+            } catch (GremlinTypeErrorException ex) {
+                if (this instanceof BinaryReductionStep || getTraversal().getParent() == EmptyStep.instance()) {
```

Comment:
@spmallette Is this the right way to test for top-level traversal?

Reply:
The reference implementation doesn't really work like that. There is nothing to forward it to. This FilterStep (if it's at the top level of the query) has the solution and makes the call whether to filter it or not. It is not embedded in a target, the solution has already left the target. If an exception is thrown here it goes straight to the client, there is no where higher to catch it inside of the query execution layer. For Graph providers that push this down obviously that changes.

Reply:
I agree with what Mike is saying. When a traversal is executed, exceptions are directly propagated outside to the server and from server to the client. One option for "guaranteed-top-level" would be catching this at server and converting to a result but I would not be in favour of that approach since it leaks the query execution semantics to the server and introduce strong coupling.

Reply:
checking for a top-level traversal is generally done with `Traversal.isRoot()`. other than that i think you have this right.